[WIP,RFT,RFC] generic: qca8k: add experimental patch for eth mdio

Ansuel opened the pull request:
Feature:
--
- Add support for mdio read/write in Ethernet packet.
- Add support for phy read/write in Ethernet packet.
- Add support for MIB counter in Ethernet packet.
- Bulk read/write. Read and write fdb entry in one go.

I need some stability testing and some perf stats.

DEPENDS
--
- Dsa conversion patch for ipq806x #4036 
- kernel 5.15 support patch #4748 

Current problem:
--
~This slow down init. I still need to find the correct way to understand when the tagger/dsa starts to receive/send packet. Because of this every request has to timeout and use the legacy mdio way.
When the system starts to handle packet from the switch port the mdio read/write will use the Ethernet way and will be fasted.~ (With the help of the kernel dev we develop a way to track when the master port (and the tagger) can correctly receive and handle packet)

Reason:
--
To support multicpu port, we probably need to enable assisted learning for cpu. This cause an increase in mdio read/write and this new way should speedup everything and remove the problem of assisted learning.

I would really appreciate any test with this.

Additions:
--
This also include another pr that permit to skip some redundant mdio write to reduce even more mdio use.
I should also test if we can use the same logic for the mdio read. In theory yes...


WARNING
--
This is not compile tested on a normal openwrt... I compiled this on linux kernel and I directly modified the source in the build dir. So expect some problem with patch apply.

I got some report that this apply directly to 5.15 pr.

@robimarko wonder if you can give some comments on the implementation if you have any idea on how this can be improved... 

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
generic: qca8k: add experimental patch for eth mdio

Add support for mdio read/write in Ethernet packet.
Add support for MIB counter in Ethernet packet.
Add support for phy read/write in Ethernet packet.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
generic: qca8k: add experimental patch to reduce mdio write

Add experimental patch to reduce mdio write by caching lo and hi part.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel commented on target/linux/generic/pending-5.10/999-5-net-dsa-tag_qca-Add-support-for-handling-mdio-read-w.patch:
```diff
@@ -0,0 +1,73 @@
++	header->data[0] = mdio_ethhdr->mdio_data;
++
++	/* Get the rest of the 12 byte of data */
++	memcpy(header->data + 1, skb->data, QCA_HDR_MDIO_DATA2_LEN);
```
I can check the cmd and skip this memcpy and the other set with write operation.

Ansuel commented on target/linux/generic/pending-5.10/999-6-net-dsa-qca8k-cache-lo-and-hi-for-mdio-write.patch:
```diff
@@ -0,0 +1,85 @@
+ 	*page = regaddr & 0x3ff;
+ }
+ 
++static u16 qca8k_current_lo = 0xffff;
```
We should consider a more safe value.... But considering the first thing will be checking the switch id... this should be fine...

rbpp commented:
@Ansuel : Hi Ansuel. Is this PR based on dsa or swconfig? I'm willing to test it but I need to know what config to use.

Ansuel commented:
it's on dsa

Il giorno ven 10 dic 2021 alle ore 22:13 rbpp ***@***.***> ha
scritto:

> @Ansuel <https://github.com/Ansuel> : Hi Ansuel. Is this PR based on dsa
> or swconfig? I'm willing to test it but I need to know what config to use.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/4828#issuecomment-991297208>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQTBNJEEJVKSUHBSMGLUQJUQRANCNFSM5JQHBDUA>
> .
>

rbpp commented:
> it's on dsa.

My question was a bit foolish, qca8k has to be on dsa... Anyway, since this PR does not contain the other required changes to work on dsa (e.g., dts changes and maybe others), do I have to apply this PR together with another one of your open PRs?

Ansuel commented:
Yes sorry I rushed this and will change shortly as I'm stabilizing it with the kernel guy.
Anyway this require 5.15 and ipq806x patch 
But if you want to wait I will push a totally different series that should also works better and have more features.
Also notice that 5.15 require firewall 4.

rbpp commented:
Take your time, I'll wait for your magic!

Ansuel commented:
I pushed the """""""final""""""""" version of this... Hope someone can test this and give some result of stability / perf.
(this is the full conversion to eth packet
this use eth packet for
- mib counter
- mdio read/write
- phy read/write
)

e9hack commented on target/linux/generic/pending-5.10/999-15-net-dsa-qca8k-cache-lo-and-hi-for-mdio-write.patch:
```diff
@@ -0,0 +1,86 @@
+ 	*page = regaddr & 0x3ff;
+ }
+ 
++static u16 qca8k_current_lo = 0xffff;
```
qca8k_current_lo/high/page must be part of private data in structure qca8k_priv. If multiple qca8k switches does exist, currently they will share this three cached values.

Ansuel replied:
Aside from the the global values that just got fixed, (still needs to be merged but i'm spending some time clearing the pr)

Caching the values looks wrong IMHO and this caching is on the regs level so we don't cache values, we cache the indirect regs used for mdio.

rbpp commented:
@Ansuel : just loaded this PR on my dumb C2600 and it's working (it's with fw3, but I don't use firewall on this AP). Any test you want me to carry?

I can also test this PR on my main R7800 but I did not figure out yet how to compile a build with firewall4 in it.

Ansuel commented:
You should check port stability and CPU load with traffic. What I need to
know is if the switch stability is equal.

Il Dom 12 Dic 2021, 15:07 rbpp ***@***.***> ha scritto:

> @Ansuel <https://github.com/Ansuel> : just loaded this PR on my dumb
> C2600 and it's working (it's with fw3, but I don't use firewall on this
> AP). Any test you want me to carry?
>
> I can also test this PR on my main R7800 but I did not figure out yet how
> to compile a build with firewall4 in it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/4828#issuecomment-991904761>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQQJ255IPV65SXC7QG3UQSUCPANCNFSM5JQHBDUA>
> .
>

rbpp commented:
I upgraded my R7800 with this PR and it's operational (with fw3). Bufferbloat test was not very good though but I can't tell if it's caused by this PR or by kernel 5.15.

Ansuel commented:
@rbpp do you have any stats? bufferbloat probably doesn't work as no firewall.
port dropping anything?

rbpp commented:
> @rbpp do you have any stats? bufferbloat probably doesn't work as no firewall. port dropping anything?

Ports were stable, no drops in the logs. My C2600 was 3 days with this PR and everything looked good. As with my R7800, I used this PR only for a day because I did not want to have my main router exposed because of the problematic firewall, but it was otherwise stable during that period.

Ansuel committed:
generic: backport patch for tagger private data

DSA tagger can now keep private data in the dsa switch. Backport
these patch.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel commented:
@rbpp was thinking, did you moved the patch to 5.15 dir or just applied this pr? Anyway the logic was wrong in the old series... I fixed it and make this more stable. Test it whatever you can :D

rbpp commented:
> @rbpp was thinking, did you moved the patch to 5.15 dir or just applied this pr?

I had applied this PR + #4748 (k5.15) + #4036 (ipq806x dsa), as per your instruction. I'll test this new version in the next few days.

KA2107 commented:
@Ansuel Thank you for your work. Just one minor thing, "experimental" spelling is wrong in both the Pull Request title and one of the commits.

Ansuel commented:
@KA2107 just notice that. Anyway I'm rushing this to make it at least compliable... I need to finish a feature and I should really start refreshing all these pr... (example the add experimental patch got merged upstream and I have to backport them to 5.15)

Ansuel committed:
generic: qca8k: add experiemntal patch for bulk read/write

mgmt Ethernet can read/write up to 16byte in one go. Add support
for this with a new helper.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

rmandrad commented:
@Ansuel - are you using netfilter on 5.15 ? nf_log_common doesn't exist  on 5.15 anymore but can't find anything on your PR that removes the dependencies
thank you

Ansuel commented:
Anyway got this merged upstream... Will close this and include backport patches in the 5.15 pr

robimarko commented:
Damn, completely missed the upstream discussion lately.

Nice job.

Ansuel commented:
@robimarko i'm still very tempted to propose the code split... wonder if they will accepted it anyway even if there isn't a user

robimarko commented:
No way it will get accepted without a user

Ansuel commented:
Anyway I feel like this should be improved... I need to understand all the build_skb logic as it looks wrong that space is allocated and free multiple time for a packet of the same size...

robimarko commented:
I wouldnt worry too much, as long as its freed and not leaked its fine

Ansuel commented:
it's problematic for ath79 devices as the cpu is not that fast

robimarko commented:
It cant be that bad, its not like you have millions of packets per second for management.

Ansuel commented:
@robimarko we have tons of packet for the state machine to check the port status 100 a seconds. but i assume they can be decreased by turning off polling.. it was enabled as the cpu port can be set to base-x but never understood why they didn't just limit it to that particular mode... But I have to investigate it... could be that polling mode can be set only on switch setup and not at the mac config call
